### PR TITLE
8332842: Optimize empty CopyOnWriteArrayList allocations

### PR DESCRIPTION
**Pull request by jengebr**
Improve `java/util/concurrent/CopyOnWriteArrayList` by eliminating needless cloning of Object[0] instances. This cloning is intended to prevent callers from changing array contents, but many `CopyOnWriteArrayList`s are allocated to size zero, or are otherwise maintained empty, so cloning is unnecessary.

Results from the included JMH benchmark:
Before:
```
Benchmark                                                    Mode  Cnt   Score   Error  Units
CopyOnWriteArrayListBenchmark.clear                          avgt    5  74.487 ± 1.793  ns/op
CopyOnWriteArrayListBenchmark.clearEmpty                     avgt    5  27.918 ± 0.759  ns/op
CopyOnWriteArrayListBenchmark.createInstanceArray            avgt    5  16.656 ± 0.375  ns/op
CopyOnWriteArrayListBenchmark.createInstanceArrayEmpty       avgt    5  15.415 ± 0.489  ns/op
CopyOnWriteArrayListBenchmark.createInstanceCollection       avgt    5  21.608 ± 0.363  ns/op
CopyOnWriteArrayListBenchmark.createInstanceCollectionEmpty  avgt    5  15.374 ± 0.260  ns/op
CopyOnWriteArrayListBenchmark.createInstanceDefault          avgt    5  15.688 ± 0.350  ns/op
CopyOnWriteArrayListBenchmark.readInstance                   avgt   10  2625.125 ± 71.802  ns/op
CopyOnWriteArrayListBenchmark.readInstanceEmpty              avgt   10  2607.447 ± 46.400  ns/op
```

After:
```
Benchmark                                                    Mode  Cnt   Score   Error  Units
CopyOnWriteArrayListBenchmark.clear                          avgt    5  75.365 ± 2.092  ns/op
CopyOnWriteArrayListBenchmark.clearEmpty                     avgt    5  20.803 ± 0.539  ns/op
CopyOnWriteArrayListBenchmark.createInstanceArray            avgt    5  16.808 ± 0.582  ns/op
CopyOnWriteArrayListBenchmark.createInstanceArrayEmpty       avgt    5  12.980 ± 0.418  ns/op
CopyOnWriteArrayListBenchmark.createInstanceCollection       avgt    5  21.627 ± 0.173  ns/op
CopyOnWriteArrayListBenchmark.createInstanceCollectionEmpty  avgt    5  12.864 ± 0.408  ns/op
CopyOnWriteArrayListBenchmark.createInstanceDefault          avgt    5  12.931 ± 0.255  ns/op
CopyOnWriteArrayListBenchmark.readInstance                   avgt   10  2615.500 ± 30.771  ns/op
CopyOnWriteArrayListBenchmark.readInstanceEmpty              avgt   10  2583.892 ± 62.086  ns/op
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8332842](https://bugs.openjdk.org/browse/JDK-8332842): Optimize empty CopyOnWriteArrayList allocations (**Enhancement** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19527/head:pull/19527` \
`$ git checkout pull/19527`

Update a local copy of the PR: \
`$ git checkout pull/19527` \
`$ git pull https://git.openjdk.org/jdk.git pull/19527/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19527`

View PR using the GUI difftool: \
`$ git pr show -t 19527`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19527.diff">https://git.openjdk.org/jdk/pull/19527.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19527#issuecomment-2145919237)

### CODE REVIEW

File: src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java
Comment (shipilev) [inline src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java:338]: Unfortunately, I think this is against the spec for this method, which explicitly says the method must allocate a new array. Yes, this change would be within the spirit of the spec ("You can modify the array", which you cannot if the object is empty), but is against the letter of it.

**Review by shipilev — COMMENTED** [review]
Some initial comments. @DougLea might want to give it a sanity check.

File: src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java
Comment (jengebr) [reply src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java:338]: I think you're right, I'll remove the change to toArray().

**Comment by shipilev** [comment]
Note there is a jcheck failure due to whitespace issues. Plus, I think the PR should still be named "8332842: Optimize empty CopyOnWriteArrayList allocations"?

**Comment by DougLea** [comment]
The jmh benchmark checks only the empty case, you need to also show lack of impact on non-empty cases.

Assuming you demonstrate this, it seems basically OK, (Deja vu previous cases including hash maps). It is only a small band-aid -- programs generating lots of them still have to allocate the COWAL object, so the savings are small compared to not generating them at all unless needed.

**Comment by jengebr** [comment]
> The jmh benchmark checks only the empty case, you need to also show lack of impact on non-empty cases.

After removing `toArray`, I think this is covered.  Can you please confirm?

> Assuming you demonstrate this, it seems basically OK, (Deja vu previous cases including hash maps). It is only a small band-aid -- programs generating lots of them still have to allocate the COWAL object, so the savings are small compared to not generating them at all unless needed.

The motivation for this change is an app that retains 23K instances of CopyOnWriteArrayList, 22K of which have zero items.  The 22K instances each have their own copy of an Object[0].  This change eliminates 343KB of retained memory and 22K objects.  I had initially modified `toArray()` for the sake of completeness but don't directly benefit from it, so I'm willing to drop that to avoid within spec.

FWIW the issue arises inside `org.jboss.resteasy` so the benefit is broad, albeit shallow.

File: src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java
Comment (viktorklang-ora) [inline src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java:103]: Looks like readObject() ought to also be changed to use the EMPTY_ELEMENTDATA if the length of the read array is 0? And perhaps `clone()`.

Comment (DougLea) [reply src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java:103]: Agreed. You might as well be consistent with this so that no one will need to re-address if some application does zillions of clones of empty lists.

Comment (jengebr) [reply src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java:103]: I've updated `readObject` to match - thanks!  `clone()` isn't a big win since the array is allocated elsewhere and the application doesn't use that code path, so I'll leave that alone for now.

Comment (jengebr) [reply src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java:103]: Sorry, I crossed wires with @DougLea 's comment, will add `clone`.

Comment (jengebr) [reply src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java:103]: `clone()` performs a shallow copy, so there is currently no `Object[]` allocated and therefore nothing to optimize.  I don't think there's any work to do here.

@DougLea @viktorklang-ora can you confirm?

**Review by shipilev — APPROVED** [review]
I think this is good. I'll let Doug and Viktor to confirm their comments were addressed.

/reviewers 2

File: src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java
Comment (viktorklang-ora) [reply src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java:103]: :+1:

**Comment by viktorklang-ora** [comment]
I think this one looks OK now. @AlanBateman might want to provide his take on this.

**Comment by AlanBateman** [comment]
> @AlanBateman -- could you please take a look? Thanks.

There was a lot of heap analysis done a few years ago that shined a light on the number of empty collections in a typical heap. I don't recall seeing COWAL in any of the data but it seems plausible that there are cases where there are a lot of empty COWAL instances in the heap.

Main thing for a change like this to make sure it doesn't hurt other cases and check if there are overridable methods used in the existing + updated implementations that might get reported as a behavior change by something that extends and overrides some methods. I don't see anything so I think it's okay.

One thing that surprises me is the change to remove(Object) to handle the case where the last remaining element is removed. Does that help the application that prompted this changed? Part of wonders if remove(int) needs to do the same as someone will show up sometime to ask.

**Comment by jengebr** [comment]
Thank you!  I've expanded coverage to remove(int) as well.

> One thing that surprises me is the change to remove(Object) to handle the case where the last remaining element is removed. Does that help the application that prompted this change? Part of me wonders if remove(int) needs to do the same as someone will show up sometime to ask.

The application benefits most from the lower footprint of the empty constructors, but I tried to expand the optimization throughout the class.  Clearly I missed a spot, thank you for catching it.

**Review by shipilev — APPROVED** [review]
Still good.

**Comment by jengebr** [comment]
@AlanBateman can you please confirm remove() is updated as intended, and approve if ready?

**Comment by jengebr** [comment]
Still waiting for reviews.

**Comment by jengebr** [comment]
@AlanBateman review please.

**Comment by AlanBateman** [comment]
> @AlanBateman review please.

Yes, I think this is okay.

**Review by AlanBateman — APPROVED** [review]

**Comment by jengebr** [comment]
Thank you!

**Review by shipilev — APPROVED** [review]
Still good.

**Pull request closed without merging**
